### PR TITLE
fix: enable vkey generation

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR enables successful verifying key generation.

### CODE REVIEW

**Comment by TomTaehoonKim** [comment]
@chokobole Can you fix typo in 56133f3? So you don't depend on it if you need a reference. -> So don't depend on it if you need a reference.

**Comment by TomTaehoonKim** [comment]
@chokobole Could you also fix typo in 99f49d8? isn't -> aren't

**Comment by chokobole** [comment]
> @chokobole
> 
> 1. Can you fix typo in [56133f3](https://github.com/kroma-network/tachyon/commit/56133f331d11c98a9106cbee713fbf44bee1b6fb)? So you don't depend on it if you need a reference. -> So don't depend on it if you need a reference.
> 2. Plus, isn't  [56133f3](https://github.com/kroma-network/tachyon/commit/56133f331d11c98a9106cbee713fbf44bee1b6fb) a perf commit?

I fixed the typo! It's a fix commit because it was one of reasons that produced a wrong result.

**Comment by TomTaehoonKim** [comment]
@chokobole Can you fix typo in 7767578? UnivaraiteEvaluation -> UnivariateEvaluation.

File: tachyon/zk/plonk/permutation/unpermuted_table.h [hunk 1]
```diff
@@ -94,9 +96,20 @@ class UnpermutedTable {
   // Calculate 𝛿 = g^2ˢ with order T (i.e., T-th root of unity),
   // where T = F::Config::kTrace.
   constexpr static F GetDelta() {
-    F g = F::FromMontgomery(F::Config::kSubgroupGenerator);
-    F adicity = F(2).Pow(F::Config::kTwoAdicity);
-    return g.Pow(adicity.ToBigInt());
+    // NOTE(chokobole): The resulting value is different from the one in
+    // https://github.com/kroma-network/halo2curves/blob/dev/src/bn256/fr.rs#L101-L110.
```
Comment (dongchangYoo) [inline tachyon/zk/plonk/permutation/unpermuted_table.h:99]: how about change this as a perm link?

File: tachyon/zk/plonk/circuit/selector_compressor.h [hunk 2]
```diff
@@ -67,14 +67,13 @@ class SelectorCompressor {
...
 
-    auto zipped = base::Zipped(selectors_in, degrees);
-    std::vector<SelectorDescription> selectors = base::Map(
-        zipped.begin(), zipped.end(),
-        [](size_t selector_index,
-           const std::tuple<std::vector<bool>, size_t>& e) {
-          const auto& [activations, max_degree] = e;
-          return SelectorDescription(selector_index, activations, max_degree);
-        });
+    std::vector<SelectorDescription> selectors;
```
Comment (dongchangYoo) [inline tachyon/zk/plonk/circuit/selector_compressor.h:70]: how about using `base::Map`?

File: tachyon/zk/plonk/circuit/selector_compressor.h [hunk 3]
```diff
@@ -67,14 +67,13 @@ class SelectorCompressor {
...
+    selectors.reserve(selectors_in.size());
+    for (size_t i = 0; i < selectors_in.size(); ++i) {
+      size_t max_degree = degrees[i];
+      const std::vector<bool>& activations = selectors_in[i];
+      selectors.push_back(SelectorDescription(i, activations, max_degree));
+    }
```
Comment (dongchangYoo) [inline tachyon/zk/plonk/circuit/selector_compressor.h]: how about using base::Map?

File: tachyon/zk/plonk/circuit/selector_compressor.h [hunk 4]
```diff
@@ -200,12 +199,12 @@ class SelectorCompressor {
         }
 
         // Update the combination assignment
-        for (auto& [combination, selector] :
-             base::Zipped(combination_assignment, selector.activations())) {
+        const std::vector<bool>& activations = selector.activations();
```
Comment (dongchangYoo) [inline tachyon/zk/plonk/circuit/selector_compressor.h:202]: Please let me know why `Zipped` was replaced.

Comment (chokobole) [reply tachyon/zk/plonk/circuit/selector_compressor.h:202]: As specified in the commit message, `Zipped` iterator copies the value. So even though I tried to modify `combination` here, it doesn't change the value in `combination_assignment`.

File: tachyon/base/strings/string_util.cc [hunk 5]
```diff
@@ -127,4 +127,14 @@ std::string MaybePrepend0x(std::string&& str) {
   return DoMaybePrepend0x(std::move(str));
 }
 
+std::string PadLeftZeroToHexString(const std::string& str, size_t num) {
```
Comment (dongchangYoo) [inline tachyon/base/strings/string_util.cc:130]: how about rename ToHexStringWithLeadingZero()?

Comment (dongchangYoo) [inline tachyon/base/strings/string_util.cc:130]: How about rename this to ToHexStringWithLeadingZero?

File: tachyon/math/polynomials/univariate/univariate_evaluations.h [hunk 6]
```diff
@@ -49,53 +49,54 @@ class UnivariateEvaluations final
...
+        base::CreateVector(degree + 1, []() { return F::Random(); }));
   }
 
   constexpr const std::vector<F>& evaluations() const { return evaluations_; }
   constexpr std::vector<F>& evaluations() { return evaluations_; }
 
+  // NOTE(chokobole): Sometimes, this degree doesn't match with the exact
+  // degree. I leave it for consistency with another polynomial.
+  // For example, [0, 0, 0, 0] gives you 3, but it's 0 in reality.
+  constexpr size_t Degree() const {
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/polynomials/univariate/univariate_evaluations.h:95]: Yeap I will

File: tachyon/math/polynomials/univariate/univariate_evaluations.h [hunk 7]
```diff
@@ -210,15 +211,16 @@ class PolynomialTraits<UnivariateEvaluations<F, MaxDegree>> {
...
+  // NOTE(chokobole): We shouldn't hash only with a non-zero term.
+  // See https://abseil.io/docs/cpp/guides/hash#the-abslhashvalue-overload
+  size_t degree = 0;
+  for (const F& eval : evals.evaluations()) {
+    h = H::combine(std::move(h), eval);
+    ++degree;
+  }
+  F zero = F::Zero();
+  for (size_t i = degree; i < MaxDegree + 1; ++i) {
+    h = H::combine(std::move(h), zero);
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/polynomials/univariate/univariate_evaluations.h:223]: If i do this, `F::Zero()` is created per every loop.

File: tachyon/zk/plonk/circuit/selector_compressor.h (hunk 4 above)
Comment (dongchangYoo) [reply tachyon/zk/plonk/circuit/selector_compressor.h:201]: Thank you for the kind explanation!

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by Insun35 — APPROVED** [review]
LGTM

**Pull request merged**